# Estudo de Caso I

## Instruções Gerais

1. Este estudo de caso será o trabalho que deverá ser entregue até o dia 28 de setembro às 23 h 59.
2. Seu valor será 6.00 e constituirá a nota da Avaliação do 1º Bimestre.
3. <ins>NÃO SERÃO ACEITAS ENTREGAS POSTERIORES À DATA LIMITE</ins>. Assim, a nota da Avaliação Bimestral será considerada zero e você deverá solicitar a Avaliação Substitutiva/2ª Chamada para recuperar essa nota.
4. Este estudo de caso poderá ser desenvolvivo e entregue individualmente ou em grupo com, no máximo, 3 integrantes.
5. Na célula abaixo insira seu nome <ins>**COMPLETO**</ins> ou o nome <ins>**COMPLETO**</ins> de cada integrante do grupo.
6. Execute este arquivo no COLAB do Google, pois alguns iten só serão executados de forma adequada naquele ambiente.
7. Após revisar o arquivo, façam o download e o envie para mim no seguinte endereço

    [marcelo.saito@unifatecpr.com.br](marcelo.saito@unifatecpr.com.br)

> Dê duplo clique nesta célula e insira o(s) nome(s) <ins>**COMPLETOS**</ins>
# André Luiz Martendal da Silva, 8ºp, Noite
# Elielton dos Santos do Counto, 8ºp, Noite
# Gustavo Teófilo Kotelak, 8ºp, Noite

# Analisando características de Smartfones

Com o intuito de analisar alguns smartfones presentes no mercado, você encontrou na internet o arquivo <ins>smartphone_cleanead_v5.csv</ins> contendo algumas informações de diversos aparelhos.

Inicialmente, você invocou a biblioteca <ins>pandas</ins> com o alias <ins>pd</ins>:

In [3]:
import pandas as pd


In [4]:
from google.colab import drive
drive.mount('/content/drive')
smartfone ='/content/drive/MyDrive/fatec/smartphone_cleaned_v5.csv'

Mounted at /content/drive


Em seguida, você leu a tabela e a atribuiu ao nome <ins>sf</ins> (de SmartFone) e visualizou as suas 5 primeiras linhas:

In [5]:
sf = pd.read_csv(smartfone)
sf.head(5)

brand_name                      model  price  rating  has_5g  has_nfc  \
0    oneplus              OnePlus 11 5G  54999    89.0    True     True   
1    oneplus  OnePlus Nord CE 2 Lite 5G  19989    81.0    True    False   
2    samsung      Samsung Galaxy A14 5G  16499    75.0    True    False   
3   motorola       Motorola Moto G62 5G  14999    81.0    True    False   
4     realme         Realme 10 Pro Plus  24999    82.0    True    False   

   has_ir_blaster processor_brand  num_cores  processor_speed  ...  \
0           False      snapdragon        8.0              3.2  ...   
1           False      snapdragon        8.0              2.2  ...   
2           False          exynos        8.0              2.4  ...   
3           False      snapdragon        8.0              2.2  ...   
4           False       dimensity        8.0              2.6  ...   

   screen_size  refresh_rate    resolution  num_rear_cameras  \
0         6.70           120  1440 x 3216                  3   
1         6.59           120  1080 x 2412                  3   
2         6.60            90  1080 x 2408                  3   
3         6.55           120  1080 x 2400                  3   
4         6.70           120  1080 x 2412                  3   

   num_front_cameras       os  primary_camera_rear primary_camera_front  \
0                1.0  android                 50.0                 16.0   
1                1.0  android                 64.0                 16.0   
2                1.0  android                 50.0                 13.0   
3                1.0  android                 50.0                 16.0   
4                1.0  android                108.0                 16.0   

   extended_memory_available  extended_upto  
0                          0            NaN  
1                          1         1024.0  
2                          1         1024.0  
3                          1         1024.0  
4                          0            NaN  

[5 rows x 25 columns]

Contudo você percebeu que havia colunas que não estavam sendo mostradas. Por esse motivo você usou

> .columns

para verificar o rótulo de cada coluna:

In [6]:
sf.columns


Index(['brand_name', 'model', 'price', 'rating', 'has_5g', 'has_nfc',
       'has_ir_blaster', 'processor_brand', 'num_cores', 'processor_speed',
       'battery_capacity', 'fast_charging_available', 'fast_charging',
       'ram_capacity', 'internal_memory', 'screen_size', 'refresh_rate',
       'resolution', 'num_rear_cameras', 'num_front_cameras', 'os',
       'primary_camera_rear', 'primary_camera_front',
       'extended_memory_available', 'extended_upto'],
      dtype='object')

Ao voltar ao site em que você achou essa tabela, você obteve as seguintes informações sobre os valores presentes em cada coluna:

1. **brand_name**: marca
2. **model**: modelo
3. **price**: preço em rupia indiana
4. **rating**: avaliações
5. **has_5g**: aceita 5G
6. **has_nfc**: apresenta NFC
7. **has_ir_blaster**: apresenta infravermelho
8. **processor_brand**: marca do processador
9. **num_cores**: quantidade de núcleos do processador
10. **processor_speed**: velocidade do processador em GHz
11. **battery_capacity**: capacidade da bateria em mAh
12. **fast_charging_available**: apresenta carregamento rápido
13. **fast_charging**: potência em W
14. **ram_capacity**: memória RAM em GB
15. **internal_memory**: memória interna em GB
16. **screen_size**: tamanho da tela em polegadas
17. **refresh_rate**: taxa de atualização da tela em Hz
18. **resolution**: resolução em pixels
19. **num_rear_cameras**: quantidade de câmeras múltiplas
20. **num_front_cameras**: quantidade de câmeras frontais
21. **os**: sistema operacional
22. **primary_camera_rear**: resolução da câmera traseira em MP
23. **primary_camera_front**: resolução ca câmera frontam em MP
24. **extended_memory_available**: possibilidade de expansão de memória em GB
25. **extended_upto**: extensão máxima da memória

Após essa breve coleta de informações, você ficou curioso em saber as classificações dos valores presentes em cada coluna e se há linhas com valores ausentes. Essa verificação você fez na célula abaixo:



In [64]:
sf.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 980 entries, 0 to 979
Data columns (total 25 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   brand_name                 980 non-null    object 
 1   model                      980 non-null    object 
 2   price                      980 non-null    int64  
 3   rating                     879 non-null    float64
 4   has_5g                     980 non-null    bool   
 5   has_nfc                    980 non-null    bool   
 6   has_ir_blaster             980 non-null    bool   
 7   processor_brand            960 non-null    object 
 8   num_cores                  974 non-null    float64
 9   processor_speed            938 non-null    float64
 10  battery_capacity           969 non-null    float64
 11  fast_charging_available    980 non-null    int64  
 12  fast_charging              769 non-null    float64
 13  ram_capacity               980 non-null    float64

Embora existam 25 colunas, é de seu interesse apenas as colunas de <ins>índices</ins>(execute a célula abaixo para saber os <ins>índices</ins> das colunas de seu interesse)

In [23]:
# Apenas execute esta célula ###################################################
import numpy as np

cols = np.sort(np.random.choice(np.arange(25), 3, replace=False))

print(f'''
Para 1ª coluna: índice {cols[0]}
Para 2ª coluna: índice {cols[1]}
Para 3ª coluna: índice {cols[2]}

Digite os índices da 1ª, 2ª e 3ª coluna que você deseja analisar nos espaços abaixo:
''')


Para 1ª coluna: índice 4
Para 2ª coluna: índice 8
Para 3ª coluna: índice 22

Digite os índices da 1ª, 2ª e 3ª coluna que você deseja analisar nos espaços abaixo:



In [ ]:
# @title
coluna1 = 4 # @param {type:"integer"}
coluna2 = 8 # @param {type:"integer"}
coluna3 = 22 # @param {type:"integer"}


No total, quantos indivíduos existem nessa tabela?

In [ ]:
# @title
total_linhas = 980 # @param {type:"integer"}


Na 1ª coluna de interesse existem quantas linhas com <ins>valores ausentes</ins>?

In [ ]:
# @title
col1_linhasVazias = 0 # @param {type:"integer"}

Na 2ª coluna de interesse existem quantas linhas com <ins>valores ausentes</ins>?

In [ ]:
# @title
col2_linhasVazias = 6 # @param {type:"integer"}

Na 3ª coluna de interesse existem quantas linhas com <ins>valores ausentes</ins>?

In [ ]:
# @title
col3_linhasVazias = 5 # @param {type:"integer"}

Depois dessas análises, você decidiu se aprofundar apenas nas colunas

* brand_name
* model
* has_5g
* has_nfc
* processor_brand
* processor_speed
* battery_capacity
* extended_upto
* price

Para isso, você criou uma <ins>cópia</ins> dessas colunas e atribuiu ao nome <ins>sf2</ins> e visualizou as 5 **últimas** colunas copiadas que geraram uma nova tabela.

In [38]:
sf2 = sf[['brand_name', 'has_5g','has_nfc','processor_brand','processor_speed','battery_capacity','extended_upto','price']]
sf2


brand_name  has_5g  has_nfc processor_brand  processor_speed  \
0      oneplus    True     True      snapdragon             3.20   
1      oneplus    True    False      snapdragon             2.20   
2      samsung    True    False          exynos             2.40   
3     motorola    True    False      snapdragon             2.20   
4       realme    True    False       dimensity             2.60   
..         ...     ...      ...             ...              ...   
975   motorola    True    False      snapdragon             3.00   
976      honor    True    False      snapdragon             2.20   
977       poco    True     True       dimensity             2.85   
978   motorola    True     True      snapdragon             2.20   
979    samsung    True    False             NaN              NaN   

     battery_capacity  extended_upto  price  
0              5000.0            NaN  54999  
1              5000.0         1024.0  19989  
2              5000.0         1024.0  16499  
3              5000.0         1024.0  14999  
4              5000.0            NaN  24999  
..                ...            ...    ...  
975            5000.0            NaN  34990  
976            5000.0         1024.0  14990  
977            5080.0            NaN  28990  
978            5000.0         1024.0  19990  
979            5000.0         1024.0  24990  

[980 rows x 8 columns]

Para facilitar o seu trabalho, você resolveu renomear as colunas da seguinte forma

* brand_name $\Rightarrow$ marca
* model $\Rightarrow$ modelo
* has_5g $\Rightarrow$ 5g
* has_nfc $\Rightarrow$ nfc
* processor_brand $\Rightarrow$ processador
* processor_speed $\Rightarrow$ velocidade(GHz)
* battery_capacity $\Rightarrow$ bateria(mAh)
* extended_upto $\Rightarrow$ expansível até(GB)
* price $\Rightarrow$ preço

Após a renomeação, visualize a 1ª linha.

In [39]:
sf2 = sf2.rename(columns={
    'brand_name': 'marca',
    'model': 'modelo',
    'has_5g': '5g',
    'has_nfc': 'nfc',
    'processor_brand': 'processador',
    'processor_speed': 'velocidade(GHz)',
    'battery_capacity': 'bateria(mAh)',
    'extended_upto': 'expansível até(GB)',
    'price': 'preço'
})
sf2.head(1)

marca    5g   nfc processador  velocidade(GHz)  bateria(mAh)  \
0  oneplus  True  True  snapdragon              3.2        5000.0   

   expansível até(GB)  preço  
0                 NaN  54999

Após fazer as renomeações, você se lembrou de que os valores da coluna *preço* estão em <ins>rupia indiana</ins>. Após realizar uma pesquisa, você chegou a um conjunto de *fatores de conversão* (fc) para algumas datas de 2023. Esses fatores estão na tabela presente no seguinte link:

[https://raw.githubusercontent.com/nightlySnoopy/database/main/convertionfactor.csv](https://raw.githubusercontent.com/nightlySnoopy/database/main/convertionfactor.csv)

atribua a tabela desse link ao nome *fatcon* e visualize suas 10 últimas linhas.

In [32]:
link = 'https://raw.githubusercontent.com/nightlySnoopy/database/main/convertionfactor.csv'
fatcon = pd.read_csv(link)
fatcon.tail(10)

data       fc
22  03/07/2023  0.05867
23  10/07/2023  0.05937
24  17/07/2023  0.05859
25  24/07/2023  0.05778
26  31/07/2023  0.05747
27  07/08/2023  0.05919
28  14/08/2023  0.05960
29  21/08/2023  0.05992
30  28/08/2023  0.05894
31  04/09/2023  0.05968

Para converter a Reais os preços, basta multiplicar o valor em rupia indiana pelo fator de conversão:

> $\text{valor em reais} = \text{valor em rupia indiana} \times \text{fator de conversão}$

Assim, se estivéssemos no dia 04 de setembro e quiséssemos converter 100 Rupias Indianas para reais, bastava fazer:

> $\text{valor em reais} = \text{valor em rupia indiana} \times \text{fator de conversão}$
>
> $\text{valor em reais} = 100 \times 0.05968$
>
> $\text{valor em reais} = 5.968 \text{ reais}$

Contudo, na tabela *fatcon* existem 30 entradas. Lembrando das aulas de Estatística Aplicada a Ciência de Dados, você resolveu calcular a média aritmética da coluna *fc* para encontrar um valor central que represente essa distribuição de valores:

In [33]:
media = fatcon['fc'].mean()
media

0.060744375

Para verificar se esses valores não estão muito "espalhados" em torno da média aritmética, você calculou a variância dessa <ins>amostra</ins>:

In [34]:
variancia = fatcon['fc'].var()
variancia

4.254844758064518e-06

e também aproveitou para calcular seu desvio padrão:

In [35]:
desvio = fatcon['fc'].std()
desvio

0.002062727504559077

Após obter esses valores da média aritmética, variância e desvio padrão amostrais você ficou na dúvida: Será que eles estão indicando que esses fatores de conversão estão concentrados próximo à média aritmética ou será que estão muito dispersos dela? Essa dúvida surgiu porque seu professor de Estatística Aplicada a Ciência de Dados não forneceu valores "padrões" para se determinar isso (uma vez que elas não existem). Mas, após você fazer uma breve pesquisa, você descobriu que sim, os valores dessas medidas indicam que os valores dos fatores de conversão estão próximos da média aritmética. Assim, você resolveu utilizar essa medida (a média aritmética) para fazer a conversão de rupia indiana para reais, ou seja:

> $\text{valor em reais} = \text{valor em rupia indiana} \times 0.060744375$

E com base nisso, você criou a coluna

> preço(R\$)

na tabela *sf2* com os valores convertidos. Em seguida você visualizou as 5 primeiras linhas dessa tabela.

In [40]:
sf2

marca    5g    nfc processador  velocidade(GHz)  bateria(mAh)  \
0     oneplus  True   True  snapdragon             3.20        5000.0   
1     oneplus  True  False  snapdragon             2.20        5000.0   
2     samsung  True  False      exynos             2.40        5000.0   
3    motorola  True  False  snapdragon             2.20        5000.0   
4      realme  True  False   dimensity             2.60        5000.0   
..        ...   ...    ...         ...              ...           ...   
975  motorola  True  False  snapdragon             3.00        5000.0   
976     honor  True  False  snapdragon             2.20        5000.0   
977      poco  True   True   dimensity             2.85        5080.0   
978  motorola  True   True  snapdragon             2.20        5000.0   
979   samsung  True  False         NaN              NaN        5000.0   

     expansível até(GB)  preço  
0                   NaN  54999  
1                1024.0  19989  
2                1024.0  16499  
3                1024.0  14999  
4                   NaN  24999  
..                  ...    ...  
975                 NaN  34990  
976              1024.0  14990  
977                 NaN  28990  
978              1024.0  19990  
979              1024.0  24990  

[980 rows x 8 columns]

In [47]:
sf2['preço(R$)'] = sf2['preço'] * media
sf2.head()

marca    5g    nfc processador  velocidade(GHz)  bateria(mAh)  \
0   oneplus  True   True  snapdragon              3.2        5000.0   
1   oneplus  True  False  snapdragon              2.2        5000.0   
2   samsung  True  False      exynos              2.4        5000.0   
3  motorola  True  False  snapdragon              2.2        5000.0   
4    realme  True  False   dimensity              2.6        5000.0   

   expansível até(GB)  preço    preço(R$)  
0                 NaN  54999  3340.879881  
1              1024.0  19989  1214.219312  
2              1024.0  16499  1002.221443  
3              1024.0  14999   911.104881  
4                 NaN  24999  1518.548631

Ao visualizar a nova coluna na tabela *sf2*, você se lembrou de seu péssimo costume de esquecer de arredondar os resultados. Assim, você atualizou a coluna *preço(R\$)* arredondando seus valores para 2 casas decimais e novamente visualizou as 5 primeiras linhas da tabela:

In [48]:
sf2['preço(R$)'] = sf2['preço(R$)'].round(2)
sf2.head()

marca    5g    nfc processador  velocidade(GHz)  bateria(mAh)  \
0   oneplus  True   True  snapdragon              3.2        5000.0   
1   oneplus  True  False  snapdragon              2.2        5000.0   
2   samsung  True  False      exynos              2.4        5000.0   
3  motorola  True  False  snapdragon              2.2        5000.0   
4    realme  True  False   dimensity              2.6        5000.0   

   expansível até(GB)  preço  preço(R$)  
0                 NaN  54999    3340.88  
1              1024.0  19989    1214.22  
2              1024.0  16499    1002.22  
3              1024.0  14999     911.10  
4                 NaN  24999    1518.55

Obtendo os preços em reais, você deletou a coluna *preço* (em Rupia Indiana) da tabela *sf2*, atualizou ela e visualizou suas 5 primeiras linhas.

In [49]:
deletar=['preço']
sf2 = sf2.drop(columns=deletar)
sf2

marca    5g    nfc processador  velocidade(GHz)  bateria(mAh)  \
0     oneplus  True   True  snapdragon             3.20        5000.0   
1     oneplus  True  False  snapdragon             2.20        5000.0   
2     samsung  True  False      exynos             2.40        5000.0   
3    motorola  True  False  snapdragon             2.20        5000.0   
4      realme  True  False   dimensity             2.60        5000.0   
..        ...   ...    ...         ...              ...           ...   
975  motorola  True  False  snapdragon             3.00        5000.0   
976     honor  True  False  snapdragon             2.20        5000.0   
977      poco  True   True   dimensity             2.85        5080.0   
978  motorola  True   True  snapdragon             2.20        5000.0   
979   samsung  True  False         NaN              NaN        5000.0   

     expansível até(GB)  preço(R$)  
0                   NaN    3340.88  
1                1024.0    1214.22  
2                1024.0    1002.22  
3                1024.0     911.10  
4                   NaN    1518.55  
..                  ...        ...  
975                 NaN    2125.45  
976              1024.0     910.56  
977                 NaN    1760.98  
978              1024.0    1214.28  
979              1024.0    1518.00  

[980 rows x 8 columns]

Agora que a tabela está mais organizada, você vai iniciar algumas análises. Execute a célula a seguir para saber com quais colunas você vai trabalhar:

In [54]:
# Apenas execute esta célula. ##################################################
import numpy as np
analysis = sf2.columns.to_numpy()
columns = np.sort(np.random.choice(analysis, 2, replace=False))

print(f'''
Você irá trabalhar com as colunas:
1ª coluna: {columns[0]}
2ª coluna: {columns[1]}

Abaixo, digite (cole) o nome das colunas que você irá analisar.
''')
col1 = columns[0]
col2 = columns[1]


Você irá trabalhar com as colunas:
1ª coluna: expansível até(GB)
2ª coluna: velocidade(GHz)

Abaixo, digite (cole) o nome das colunas que você irá analisar.



In [55]:
# @title
col1 = "expans\xEDvel at\xE9(GB)" # @param {type:"string"}
col1 = col1.strip()


In [56]:
# @title
col2 = "velocidade(GHz)" # @param {type:"string"}
col2 = col2.strip()


Execute a célula abaixo para saber quais informações das colunas serão necessárias utilizar para realizar a filtragem das linhas de interesse.

In [57]:
# Execute esta célula. Não altere as informações contidas nela. ################
def tobe():
    return np.random.choice(['é', 'não é'])

def tobe2():
    return np.random.choice(['seja', 'não seja'])

def andor():
    return np.random.choice(['e', 'ou'])

def hasnot():
    return np.random.choice(['tem', 'tão tem'])

def ineq1():
    l = ['maior que', 'menor que', 'maior ou igual a',
         'menor ou igual a', 'igual a']
    return np.random.choice(l)

def ineq2():
    l = ['mais que', 'menos que', 'mais que ou igual a',
         'menos que ou igual a', 'igual a']
    return np.random.choice(l)
################################################################################
if col1 == 'marca':
    brand = np.random.choice(sf2['marca'].unique())
    condm = tobe()
    connector = andor()
    text1 = f'A marca {condm} {brand} {connector} '

elif col1 == 'modelo':
    model = np.random.choice(sf2['modelo'].unique())
    condm = tobe()
    connector = andor()
    text1 = f'O modelo {condm} {model} {connector}'

elif col1 == '5g':
    condm = tobe().capitalize()
    connector = andor()
    text1 = f'{condm} 5g {connector} '

elif col1 == 'nfc':
    condm = hasnot().capitalize()
    connector = andor()
    text1 = f'{condm} NFC {connector} '

elif col1 == 'processador':
    condm = np.random.choice(sf2['processador'].unique())
    condition = tobe()
    connector = andor()
    text1 = f'O processador {condition} {condm} {connector} '

elif col1 == 'velocidade(GHz)':
    condition = ineq1()
    value = np.round(np.random.randint(1200, 3221) / 1000, 2)
    connector = andor()
    text1 = f'A velocidade é {condition} {value} GHz {connector} '

elif col1 == 'bateria(mAh)':
    condition = ineq1()
    value = np.random.choice(sf2['bateria(mAh)'].unique())
    connector = andor()
    text1 = f'A bateria tenha capacidade {condition} {value} mAh {connector} '

elif col1 == 'expansível até(GB)':
    value = 2 ** np.random.choice(np.arange(6, 12))
    connector = andor()
    text1 = f'Seja expansível até {value} GB {connector} '

else:
    condition = ineq2()
    value = np.random.choice(sf2['preço(R$)'].unique())
    connector = andor()
    text1 = f'O valor seja {condition} {value} reais {connector} '

if col2 == 'marca':
    brand = np.random.choice(sf2['marca'].unique())
    condm = tobe2()
    text2 = f'a marca {condm} {brand}.'

elif col2 == 'modelo':
    model = np.random.choice(sf2['modelo'].unique())
    condm = tobe2()
    text2 = f'o modelo {condm} {model}.'

elif col2 == '5g':
    condm = tobe2()
    text2 = f'{condm} 5g.'

elif col2 == 'nfc':
    condm = hasnot()
    text2 = f'{condm} NFC.'

elif col2 == 'processador':
    condm = np.random.choice(sf2['processador'].unique())
    condition = tobe()
    text2 = f'o processador {condition} {condm}.'

elif col2 == 'velocidade(GHz)':
    condition = ineq1()
    value = np.round(np.random.randint(1200, 3221) / 1000, 2)
    text2 = f'a velocidade seja {condition} {value} GHz. '

elif col2 == 'bateria(mAh)':
    condition = ineq1()
    value = np.random.choice(sf2['bateria(mAh)'].unique())
    text2 = f'a bateria tenha capacidade {condition} {value} mAh.'

elif col2 == 'expansível até(GB)':
    value = np.random.choice([64, 128, 256, 512, 1024, 2048])
    text2 = f'seja expansível até {value} GB.'

else:
    condition = ineq2()
    value = np.random.choice(sf2['preço(R$)'].unique())
    text2 = f'O valor seja {condition} {value} reais.'

print(text1 + text2)
print('''
Copie o texto acima e cole-o na célula abaixo.

(Atenção: Cada vez que você executar esta célula, valores distintos serão gerados.)
''')

Seja expansível até 256 GB e a velocidade seja igual a 1.42 GHz. 

Copie o texto acima e cole-o na célula abaixo.

(Atenção: Cada vez que você executar esta célula, valores distintos serão gerados.)



In [58]:
# @title
filtro = "Seja expans\xEDvel at\xE9 256 GB e a velocidade seja igual a 1.42 GHz" # @param {type:"string"}


Assim, fazendo a filtragem:

In [60]:
filtro = sf2[(sf2['expansível até(GB)'] <= 256 ) & (sf2['velocidade(GHz)'] == 1.42)]
filtro.shape[0]

0

A quantidade de aparelhos que apresentam as características analisadas foram

In [61]:
# @title
aparelhos = 0 # @param{type: 'integer'}
